native/java/org/jpype/ref/JPypeReferenceSet.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+package org.jpype.ref;
+
+import java.util.ArrayList;
+import static org.jpype.ref.JPypeReferenceQueue.removeHostReference;
+
+/**
+ *
+ * @author nelson85
+ */
+public class JPypeReferenceSet
+{
+
+  static final int SIZE = 256;
+  ArrayList<Pool> pools = new ArrayList<>();
+  Pool current;
+  long context;
+  private int items;
+
+  JPypeReferenceSet(long context)
+  {
+    this.context = context;
+  }
+
+  int size()
+  {
+    return items;
+  }
+
+  /**
+   * Add a reference to the set.
+   *
+   * This should be O(1).
+   *
+   * @param ref
+   */
+  synchronized void add(JPypeReference ref)
+  {
+    if (ref.cleanup == 0)
+      return;
+
+    this.items++;
+    if (current == null)
+    {
+      current = new Pool(pools.size());
+      pools.add(current);
+    }
+
+    if (current.add(ref))
+    {
+      // It is full
+      current = null;
+
+      // Find a free pool
+      for (Pool pool : pools)
+      {
+        if (pool.tail < SIZE)
+        {
+          current = pool;
+          return;
+        }
+      }
+    }
+  }
+
+  /**
+   * Remove a reference from the set.
+   *
+   * @param ref
+   */
+  synchronized void remove(JPypeReference ref)
+  {
+    if (ref.cleanup == 0)
+      return;
+    pools.get(ref.pool).remove(ref);
+    this.items--;
+    ref.cleanup = 0;
+    ref.pool = -1;
+  }
+
+  /**
+   * Release all resources.
+   *
+   * This is triggered by shutdown to release an current Python references that
+   * a being held.
+   *
+   */
+  void flush()
+  {
+    for (Pool pool : pools)
+    {
+      for (int i = 0; i < pool.tail; ++i)
+      {
+        JPypeReference ref = pool.entries[i];
+        long hostRef = ref.hostReference;
+        long cleanup = ref.cleanup;
+        if (cleanup == 0)
+        {
+          System.out.println("Fault 2");
```

Comment:
if you must print, I would prefer System.err

Reply:
Naw that was just a missed debugging line on an error on a sanity check for an error during development.  Removing and commenting appropriately.